File: scripts/lib_js/niche-firestore-sync.js
// Synchronise config/niches/<id>/niche.json avec Firestore (collection
// `niches`, un document par niche = le JSON entier, largement sous la limite
// de 1 Mo/document — voir plan MVP SaaS interne).
//
// Design choisi pour éviter un refactor async de tout le pipeline : Firestore
// devient la source canonique, mais niche-config.js / niche-paths.js /
// persona.js / config.js restent des lecteurs SYNCHRONES du fichier local
// (aucun changement requis dedans). Le pont entre les deux est un `pull`
// explicite, invoqué en étape PRÉALABLE (workflow CI, ou avant un run manuel)
// — jamais depuis l'intérieur de run.js lui-même, où tout le graphe de
// require (config.js, persona.js) s'exécute avant qu'un premier `await`
// puisse avoir eu lieu.
//
// Repli gracieux : si FIREBASE_SERVICE_ACCOUNT_JSON n'est pas configuré,
// pull()/push() se contentent d'un avertissement — le fichier local reste
// la seule source de vérité, comme avant cette migration (dev local sans
// Firebase configuré toujours fonctionnel).
const fs = require('fs');
const path = require('path');
const nicheConfig = require('./niche-config');

function hasFirestoreConfigured() {
  return Boolean(process.env.FIREBASE_SERVICE_ACCOUNT_JSON) || (() => {
    try {
      const envPath = path.join(__dirname, '..', '..', '.env');
      return fs.existsSync(envPath) && fs.readFileSync(envPath, 'utf8').includes('FIREBASE_SERVICE_ACCOUNT_JSON');
    } catch {
      return false;
    }
  })();
}

/** Firestore -> fichier local. Écrase config/niches/<id>/niche.json avec le document distant. */
async function pullNiche(nicheId) {
  if (!hasFirestoreConfigured()) {
    console.warn(`niche-firestore-sync: FIREBASE_SERVICE_ACCOUNT_JSON absent — pull ignoré, fichier local conservé tel quel (${nicheId}).`);
    return { skipped: true };
  }
  const { getDb } = require('./firestore-client');
  const snap = await getDb().collection('niches').doc(nicheId).get();
  if (!snap.exists) {
    throw new Error(`niche-firestore-sync: document Firestore niches/${nicheId} introuvable — utiliser push d'abord pour l'y créer.`);
  }
  const niche = snap.data();
  const filePath = path.join(nicheConfig.nicheDir(nicheId), 'niche.json');
  fs.writeFileSync(filePath, JSON.stringify(niche, null, 2) + '\n', 'utf8');
  return { skipped: false, niche };
}

/** Fichier local -> Firestore. Le fichier local reste la source de vérité au moment du push. */
async function pushNiche(nicheId) {
  if (!hasFirestoreConfigured()) {
    console.warn(`niche-firestore-sync: FIREBASE_SERVICE_ACCOUNT_JSON absent — push ignoré (${nicheId}).`);
    return { skipped: true };
  }
  const niche = nicheConfig.loadNiche(nicheId);
  const { getDb } = require('./firestore-client');
  await getDb().collection('niches').doc(nicheId).set(niche);
  return { skipped: false, niche };
}

module.exports = { pullNiche, pushNiche, hasFirestoreConfigured };

File: scripts/lib_py/niche_firestore_sync.py
# Miroir Python de scripts/lib_js/niche-firestore-sync.js — pousse
# config/niches/<id>/niche.json vers Firestore (niches/{id}) après une
# création/édition locale (new-niche.py, add-silo.py). Repli gracieux : si
# FIREBASE_SERVICE_ACCOUNT_JSON n'est pas configuré, avertit et ne bloque
# jamais le script appelant — le fichier local reste utilisable seul, comme
# avant cette migration.
import json
import os

import niche_config

try:
    import firestore_client
except ImportError:
    firestore_client = None


def _has_firestore_configured():
    if os.environ.get('FIREBASE_SERVICE_ACCOUNT_JSON'):
        return True
    env_path = os.path.join(niche_config.ROOT, '.env')
    if os.path.exists(env_path):
        with open(env_path, encoding='utf-8') as f:
            return 'FIREBASE_SERVICE_ACCOUNT_JSON' in f.read()
    return False


def push_niche(niche_id):
    """Fichier local -> Firestore. Best-effort : n'échoue jamais le script appelant."""
    if not _has_firestore_configured() or firestore_client is None:
        print(f"niche_firestore_sync: FIREBASE_SERVICE_ACCOUNT_JSON absent — push ignore ({niche_id}).")
        return
    niche = niche_config.load_niche(niche_id)
    db = firestore_client.get_db()
    db.collection('niches').document(niche_id).set(niche)
    print(f"niche_firestore_sync: niche.json -> Firestore niches/{niche_id} : ok")
